### PR TITLE
syntax: integrate `TokenStream`

### PR DESCRIPTION
**Pull request by jseyfried**
Use `TokenStream` instead of `Vec<TokenTree>` in `TokenTree::Delimited` and elsewhere.
r? @nrc

### CODE REVIEW

**Comment by jseyfried** [comment]
cc #38356 
cc @abonander @keeperofdakeys

File: src/libsyntax/util/parser_testing.rs [hunk 1]
```diff
@@ -9,17 +9,17 @@
...
-use tokenstream;
+use tokenstream::TokenStream;
 use std::iter::Peekable;
 
 /// Map a string to tts, using a made-up filename:
-pub fn string_to_tts(source_str: String) -> Vec<tokenstream::TokenTree> {
+pub fn string_to_stream(source_str: String) -> TokenStream {
     let ps = ParseSess::new();
-    filemap_to_tts(&ps, ps.codemap().new_filemap("bogofile".to_string(), None, source_str))
+    filemap_to_stream(&ps, ps.codemap().new_filemap("bogofile".to_string(), None, source_str))
```
Comment (abonander) [inline src/libsyntax/util/parser_testing.rs]: Will this filename appear anywhere? If so, it might be a good idea to use something more meaningful.

Comment (jseyfried) [inline src/libsyntax/util/parser_testing.rs]: No -- this is only used in internal unit tests.

**Review by nrc — APPROVED** [review]
r=me with the renaming and comments

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -175,12 +175,107 @@ pub struct Parser<'a> {
...
     pub root_module_name: Option<String>,
     pub expected_tokens: Vec<TokenType>,
-    pub tts: Vec<(TokenTree, usize)>,
+    cursor: Cursor,
     pub desugar_doc_comments: bool,
     /// Whether we should configure out of line modules as we parse.
     pub cfg_mods: bool,
 }
 
+struct Cursor {
```
Comment (jseyfried) [inline src/libsyntax/parse/parser.rs]: Yeah, `TokenCursor` sounds good.

File: src/libsyntax/tokenstream.rs [hunk 3]
```diff
@@ -396,6 +328,33 @@ impl Cursor {
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct ThinTokenStream(Option<RcSlice<TokenStream>>);
```
Comment (jseyfried) [inline src/libsyntax/tokenstream.rs]: Will do.

**Comment by alexcrichton** [comment]
@bors: retry

* #38878 osx linker segfault

**Pull request merged**